src/common/commonutils/UrlUtils.c

```diff
@@ -3,24 +3,26 @@
 
 #include "Internal.h"
 
-char* UrlEncode(char* target)
+char* UrlEncode(const char* target, size_t targetSize)
 {
-    if (NULL == target)
+    size_t i = 0, j = 0;
+    int encodedLength = 0;
+    char* encodedTarget = NULL;
+
+    if ((NULL == target) || (0 == targetSize))
     {
         return NULL;
     }
 
-    int i = 0, j = 0;
-    int targetLength = (int)strlen(target);
-    int encodedLength = (3 * targetLength) + 1;
-    char* encodedTarget = (char*)malloc(encodedLength);
-    if (NULL != encodedTarget)
+    encodedLength = (3 * targetSize) + 3;

```

Comment:
Why +3?

Reply:
(targetSize + 1) * 3? Don't remember why added, won't remove that now and risk breaking the old IoT Hub telemetry this is used for.

Reply:
Decided to refactor. Please see last iteration.